invenio_drafts_resources/services/draft.py

```diff
@@ -71,85 +62,138 @@ def draft_cls(self):
 
-        return self.config.resource_unit_cls(pid=pid, record=draft)
+        return self.resource_unit(pid=pid, record=draft, links=None)
+
+    def _patch_data(self, record, data):
+        """Temporarily here until the merge strategy is set."""
+        # TODO: Implement better update/merge strategy
+        for key in data.keys():
+            record[key] = data[key]
 
     def edit(self, id_, data, identity):
-        """Create a draft for an existing record.
+        """Create a new revision or a draft for an existing record.
 
         :param id_: record PID value.
         """
-        pid, record = self.resolve(id_)
+        pid, record = self.pid_manager.resolve(id_)
         self.require_permission(identity, "create")
-        validated_data = self.draft_data_validator.validate(data)
-        draft = self.draft_cls.create(validated_data, id_=record.id,
+        validated_data = self.data_validator.validate(data, partial=True)
+        self._patch_data(record, validated_data)
+        draft = self.draft_cls.create(record.dumps(), id_=record.id,
                                       fork_version_id=record.revision_id)
         db.session.commit()  # Persist DB
         if self.indexer:
             self.indexer.index(draft)
 
-        return self.config.resource_unit_cls(pid=pid, record=draft)
+        return self.resource_unit(pid=pid, record=draft, links=None)
+
+    def _publish_revision(self, validated_data, pid):
+        """Publish draft of existing record (edition)."""
+        pid, record = self.pid_manager.resolve(pid.pid_value)
+        record.clear()
+        record.update(validated_data)
+        record.commit()
+
+        return record
+
+    def _publish_new_version(self, validated_data, draft):
+        """Publish draft of a new record."""
+        record = self.record_cls.create(validated_data, id_=draft.id)
+        conceptrecid = record['conceptrecid']
+
+        # FIXME: This implies two db queries for the same in pid_manager
+        if self.pid_manager.is_first_version(
+            pid_value=conceptrecid
+        ):
+            self.pid_manager.register_pid(record['conceptrecid'])
+
+        self.pid_manager.register_pid(record['recid'])
+
+        return record
 
     def publish(self, id_, identity):
         """Publish a draft."""
         self.require_permission(identity, "publish")
         # Get draft
-        pid, draft = self.resolve_draft(id_=id_)
-        # Validate and create record, register PID
-        data = draft.dumps()
+        pid, draft = self.pid_manager.resolve(id_, draft=True)
         # Validate against record schema
-        validated_data = self.data_validator.validate(data)
-        # Use same UUID
-        record = self.record_cls.create(validated_data, id_=draft.id)
-        pid.register()
+        validated_data = self.data_validator.validate(draft.dumps())
+        # Publish it
+        if self.pid_manager.is_published(pid):  # Then we publish a revision
```

Comment:
note to self: why the disctinction?

Reply:
This is mostly for the workflow. We have three cases:

- Record does not exists: We create a draft, then later it can be published. In this case *is_published == False*, then we need to register the `recid` (and `conceptrecid`). In addition, we have to create the record itself in the db.

- Record exists, we create a new version: In this case, we create a draft with a different `recid` (same `conceptrecid`). That means that when publishing the new version, the record *is_published == False*. Then we have to register (just) the `recid` and create the record itself in db.

- Record exists, we want to edit it: This means, changing metadata (what happens when `service.edit()`). This just need to update the json blob of the current record, there is no need to register pids or anything else. 

The distinction in the end comes from: `new version` happens when the files of the record change, `new revision` happens when the metadata changes. However, if the user decides to click `edit` won't be able to change files, and if clicks `new version` but doesn't change files it won't be able to publish.

Reply:
(haha I really need to do a re-read of my code review before submitting - this was meant to be removed :sweat_smile: )